File: built-in plugins dsp/mb-reverb-plate.cpp
/**
 * MB Plate Reverb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic plate reverb
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_PLATE_H
#define MB_REVERB_PLATE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbPlate : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-plate";
    static constexpr const char* PLUGIN_NAME    = "MB Plate Reverb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float size = 0.6f;  // range [0, 1]
    float decay = 2.0f;  // range [0.5, 6]
    float mix = 0.35f;  // range [0, 1]
    };

    MbReverbPlate() = default;
    ~MbReverbPlate() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.size = std::clamp(params.size, 0f, 1f);
        params.decay = std::clamp(params.decay, 0.5f, 6f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Plate Reverb
        return input;
    }
};

#endif // MB_REVERB_PLATE_H

File: built-in plugins dsp/mb-reverb-plate.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbPlatePlugin: PluginDefinition = { id: 'mb-reverb-plate', slug: 'mb-reverb-plate', name: 'MB Plate Reverb', category: 'effect', type: 'reverb', version: '1.0.0', description: 'Classic plate reverb', author: 'Max Booster', grade: 'A', parameters: [{ id: 'size', name: 'Size', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true }, { id: 'decay', name: 'Decay', type: 'float', defaultValue: 2.0, minValue: 0.5, maxValue: 6, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.35, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { size: 0.6, decay: 2.0, mix: 0.35 } };

export default MbReverbPlatePlugin;
